File: VB/SnapModule/Helpers/InplaceDocumentCacheHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports DevExpress.ExpressApp.Utils
Imports SnapModule.Base
Imports DevExpress.ExpressApp
Imports DevExpress.Data.Filtering
Imports System.Collections

Namespace SnapModule.Helpers
    Public Class InplaceDocumentsCacheHelper
        Private objectSpace As IObjectSpace
        Private documentDataType As Type
        Public Sub New(ByVal objectSpace As IObjectSpace, ByVal documentDataType As Type)
            Me.objectSpace = objectSpace
            Me.documentDataType = documentDataType
        End Sub
        Private isCompleteCache As Boolean = False
        Private inplaceDocumentsCache As New LightDictionary(Of Type, List(Of IDocumentData))()
        Public Overridable Function GetDocumentDataList(ByVal targetObjectType As Type) As List(Of IDocumentData)
            EnsureCache()
            Dim cachedDocuments As New List(Of IDocumentData)()
            For Each key As Type In inplaceDocumentsCache.Keys
                If key.IsAssignableFrom(targetObjectType) Then
                    cachedDocuments.AddRange(inplaceDocumentsCache(key))
                End If
            Next key
            Return cachedDocuments
        End Function
        Public Sub ClearInplaceDocumentsCache()
            isCompleteCache = False
            inplaceDocumentsCache.Clear()
        End Sub
        Protected Overridable Function CollectAllInplaceDocumentsData() As IList
            Return objectSpace.GetObjects(documentDataType, CriteriaOperator.Parse("InplaceDocumentDataSource is not null"))
        End Function
        Private Sub EnsureCache()
            If (Not isCompleteCache) Then
                Dim allInPlaceDocumentsData As IList = CollectAllInplaceDocumentsData()
                For Each item As IDocumentData In allInPlaceDocumentsData
                    If item.InplaceDocumentDataSource IsNot Nothing Then
                        Dim items As List(Of IDocumentData)
                        If (Not inplaceDocumentsCache.TryGetValue(item.InplaceDocumentDataSource.DataType, items)) Then
                            items = New List(Of IDocumentData)()
                            inplaceDocumentsCache(item.InplaceDocumentDataSource.DataType) = items
                        End If
                        items.Add(item)
                    End If
                Next item
                isCompleteCache = True
            End If
        End Sub
    End Class
End Namespace

File: CS/SnapModule/Helpers/InplaceDocumentCacheHelper.cs
using System;
using System.Collections.Generic;
using DevExpress.ExpressApp.Utils;
using SnapModule.Base;
using DevExpress.ExpressApp;
using DevExpress.Data.Filtering;
using System.Collections;

namespace SnapModule.Helpers {
    public class InplaceDocumentsCacheHelper {
        private IObjectSpace objectSpace;
        private Type documentDataType;
        public InplaceDocumentsCacheHelper(IObjectSpace objectSpace, Type documentDataType) {
            this.objectSpace = objectSpace;
            this.documentDataType = documentDataType;
        }
        private bool isCompleteCache = false;
        private LightDictionary<Type, List<IDocumentData>> inplaceDocumentsCache = new LightDictionary<Type, List<IDocumentData>>();
        public virtual List<IDocumentData> GetDocumentDataList(Type targetObjectType) {
            EnsureCache();
            List<IDocumentData> cachedDocuments = new List<IDocumentData>();
            foreach (Type key in inplaceDocumentsCache.Keys) {
                if (key.IsAssignableFrom(targetObjectType)) {
                    cachedDocuments.AddRange(inplaceDocumentsCache[key]);
                }
            }
            return cachedDocuments;
        }
        public void ClearInplaceDocumentsCache() {
            isCompleteCache = false;
            inplaceDocumentsCache.Clear();
        }
        protected virtual IList CollectAllInplaceDocumentsData() {
            return objectSpace.GetObjects(documentDataType, CriteriaOperator.Parse("InplaceDocumentDataSource is not null"));
        }
        private void EnsureCache() {
            if (!isCompleteCache) {
                IList allInPlaceDocumentsData = CollectAllInplaceDocumentsData();
                foreach (IDocumentData item in allInPlaceDocumentsData) {
                    if (item.InplaceDocumentDataSource != null) {
                        List<IDocumentData> items;
                        if (!inplaceDocumentsCache.TryGetValue(item.InplaceDocumentDataSource.DataType, out items)) {
                            items = new List<IDocumentData>();
                            inplaceDocumentsCache[item.InplaceDocumentDataSource.DataType] = items;
                        }
                        items.Add(item);
                    }
                }
                isCompleteCache = true;
            }
        }
    }
}
